Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go

```diff
@@ -368,253 +338,290 @@ func FindSourceController(r RollingUpdaterClient, namespace, name string) (*api.
-	if newRc.Spec.Replicas == desired {
+func (r *RollingUpdater) scaleUp(newRc, oldRc *api.ReplicationController, desired, increment int, config *RollingUpdaterConfig) (*api.ReplicationController, error) {
+	// If we're already at the desired, or if the current safe increment is 0,
+	// do nothing.
+	if newRc.Spec.Replicas == desired || increment == 0 {
 		return newRc, nil
 	}
-	newRc.Spec.Replicas += scaleAmount
-	if newRc.Spec.Replicas > desired || oldRc.Spec.Replicas == 0 {
+	// If the old is already scaled down, go ahead and scale all the way up.
+	if oldRc.Spec.Replicas == 0 {
+		increment = desired - newRc.Spec.Replicas
+	}
+	newRc.Spec.Replicas += increment
+	if newRc.Spec.Replicas > desired {
 		newRc.Spec.Replicas = desired
 	}
-	fmt.Fprintf(out, "Scaling %s up to %d\n", newRc.Name, newRc.Spec.Replicas)
-	scaledRc, err := r.scaleAndWait(newRc, retry, wait)
+	// TODO: remove debugging output
+	fmt.Printf("Scaling %s up (current=%d, desired=%d, increment=%d)\n", newRc.Name, newRc.Spec.Replicas, desired, increment)
+	fmt.Fprintf(config.Out, "Scaling %s up to %d\n", newRc.Name, newRc.Spec.Replicas)
+	retryWait := &RetryParams{config.Interval, config.Timeout}
+	scaledRc, err := r.scaleAndWait(newRc, retryWait, retryWait)
 	if err != nil {
 		return nil, err
 	}
-	err = config.UpdateAcceptor.Accept(scaledRc)
-	if err != nil {
-		return nil, fmt.Errorf("update rejected for %s: %v", scaledRc.Name, err)
-	}
 	return scaledRc, nil
 }
 
-// scaleDown scales down oldRc to 0 by scaleAmount. It accounts for fencepost
+// scaleDown scales down oldRc to 0 by increment. It accounts for fencepost
 // conditions. If oldRc is already scaled to 0, scaleDown does nothing. If
 // newRc is already scaled to desired, oldRc is scaled to 0 immediately
-// regardless of scaleAmount.
-func (r *RollingUpdater) scaleDown(newRc, oldRc *api.ReplicationController, desired, scaleAmount int, retry, wait *RetryParams, out io.Writer, config *RollingUpdaterConfig) (*api.ReplicationController, error) {
+// regardless of increment.
+func (r *RollingUpdater) scaleDown(newRc, oldRc *api.ReplicationController, desired, maxIncrement, minReady int, config *RollingUpdaterConfig) (*api.ReplicationController, int, error) {
 	if oldRc.Spec.Replicas == 0 {
-		return oldRc, nil
+		return oldRc, maxIncrement, nil
+	}
+	oldReady, newReady, ready, err := r.waitForReadyPods(config.Interval, config.Timeout, oldRc, newRc)
+	if err != nil {
+		return nil, 0, err
+	}
+	increment := int(math.Min(float64(maxIncrement), float64(ready-minReady)))
+	// TODO: remove debugging output
+	fmt.Printf("Scaling %s down (current=%d, maxIncrement=%d, ready=%d, minReady=%d, oldReady=%d, newReady=%d, increment=%d)\n", oldRc.Name, oldRc.Spec.Replicas, maxIncrement, ready, minReady, oldReady, newReady, increment)
+	// We can't scale down without crossing the minimum. Try again later.
+	if increment == 0 {
```

Comment:
I guess for now we can get away with not doing anything sophisticated about this but we should definitely know about it.